### PR TITLE
Language Detection

### PR DESCRIPTION
**Pull request by Abhinay1997**
PR for the issue #29.

* Adds `LanguageLogitsFilter` that suppresses non-language logits
* Implements `detectLanguage` in `TextDecoder` that does a single forward pass from `<startoftranscript>` to get the language token
* Adds tests for `LanguageLogitsFilter`

### CODE REVIEW

File: Sources/WhisperKit/Core/TextDecoder.swift [hunk 1]
```diff
@@ -308,14 +316,101 @@ public class TextDecoder: TextDecoding, WhisperMLModel {
...
-        options decoderOptions: DecodingOptions,
+        sampler tokenSampler: TokenSampling,
+        options: DecodingOptions,
         temperature: FloatType
-    ) -> [DecodingResult] {
+    ) async throws -> [DecodingResult] {
         // TODO: implement
         // predict logits for 1 iteration with sot
         // 1. LanguageLogitsFilter for only language tokens
         // 2. GreedyTokenSampler for most likely language
```
Comment (ZachNagengast) [inline Sources/WhisperKit/Core/TextDecoder.swift:326]:
```suggestion
        // Predict logits for 1 iteration with sot
        // 1. LanguageLogitsFilter for only language tokens
        // 2. GreedyTokenSampler for most likely language
```

**Review by ZachNagengast — COMMENTED** [review]
This is looking really good. There are some audio files already in the test bundle that you can use for spanish and japanese https://github.com/argmaxinc/WhisperKit/tree/main/Tests/WhisperKitTests/Resources It would be great to have an additional test that checks for accuracy with these similar to the other language tests https://github.com/argmaxinc/WhisperKit/blob/59cb8516c708e3e2f18198002600026b5a1135ca/Tests/WhisperKitTests/UnitTests.swift#L348

It would also be great to see some sort of performance comparison of using the filter vs not, for example you can cherry pick this commit in order to track overall filtering time https://github.com/argmaxinc/WhisperKit/pull/45/commits/ee95bb0789e277f45981b7027744958aea78eecc, ideally it is as small of an increase as possible to the overall pipeline

**Comment by ZachNagengast** [comment]
Added a couple CI fixes here FYI @Abhinay1997 should be in a working state now

**Comment by Abhinay1997** [comment]
Thanks for the feedback ! Yes, let me add tests for Spanish and Japanese. Will need to look into the filtering time !

**Comment by Abhinay1997** [comment]
@ZachNagengast, for comparing the detection with and without logits filter, I'll add a `applyFilter` param to `detectLanguage` then ? To toggle it on and off.

**Comment by Abhinay1997** [comment]
The filter adds only `2.9802322387695312e-06` seconds. Negligible imo

File: Sources/WhisperKit/Core/LogitsFilter.swift [hunk 2]
```diff
@@ -62,3 +62,38 @@ public class TimestampRulesFilter: LogitsFiltering {
...
+    
+    private func getNonLanguageTokenIndexes(logitsDim: Int, allLanguageTokens: [Int]) -> [[NSNumber]]{
+        var indexes: [[NSNumber]] = []
+        for i in 0..<logitsDim{
+            if !allLanguageTokens.contains(i){
+                indexes.append([0, 0, i as NSNumber])
+            }
+        }
+        return indexes
+    }
```
Comment (jkrukowski) [inline Sources/WhisperKit/Core/LogitsFilter.swift:98]: - maybe this can be precomputed in the class `init`?
- converting `allLanguageTokens` to set (instead of the array) might result in more perfomant code since `contains` for set should be O(1)

Comment (Abhinay1997) [reply Sources/WhisperKit/Core/LogitsFilter.swift:98]: Agreed ! Let me do that !

File: Sources/WhisperKit/Core/TextDecoder.swift [hunk 3]
```diff
@@ -308,14 +316,100 @@ public class TextDecoder: TextDecoding, WhisperMLModel {
...
+        logProbs = sampleResult.logProbs
+        
+        let samplingTime = Date().timeIntervalSince(samplingStartTime)
+        timings.decodingSampling += samplingTime
+        
+        var detectedLanguage = tokenizer.decode(tokens: [nextToken]).dropFirst(2).dropLast(2)
+        var decodingResult = DecodingResult.emptyResults
+        decodingResult.timings = timings
+        decodingResult.language = String(detectedLanguage)
+        return [decodingResult]
```
Comment (ZachNagengast) [inline Sources/WhisperKit/Core/TextDecoder.swift:411]: There's quite a bit of repeated code here that may get out of date with the main decoder loop, if you're up for it would be great to see an attempt at unifying them, but won't block merge.

File: Tests/WhisperKitTests/UnitTests.swift [hunk 4]
```diff
@@ -563,6 +563,18 @@ final class UnitTests: XCTestCase {
...
+        let tokensFilter1 = LanguageLogitsFilter(allLanguageTokens: [2, 4, 6], logitsDim: 7, sampleBegin: 0)
+        let logits1 = try MLMultiArray.logits([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7])
+        let result1 = tokensFilter1.filterLogits(logits1, withTokens: [])
+        XCTAssertEqual(result1.data(for: 2), [-.infinity, -.infinity, 0.3, -.infinity, 0.5, -.infinity, 0.7])
+        
+        let tokensFilter2 = LanguageLogitsFilter(allLanguageTokens: [2, 4, 6], logitsDim: 7, sampleBegin: 0)
+        let logits2 = try MLMultiArray.logits([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7])
+        let result2 = tokensFilter2.filterLogits(logits2, withTokens: [1])
+        XCTAssertEqual(result2.data(for: 2), [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7])
+    }
```
Comment (ZachNagengast) [inline Tests/WhisperKitTests/UnitTests.swift:577]: Should test language detection using the full tiny pipeline tests here for Spanish and Japanese, similar to these tests https://github.com/argmaxinc/WhisperKit/blob/main/Tests/WhisperKitTests/UnitTests.swift#L348

File: Sources/WhisperKit/Core/WhisperKit.swift [hunk 5]
```diff
@@ -646,12 +646,25 @@ public class WhisperKit: Transcriber {
...
+                    ).first
+                    currentDecodingOptions.language = languageDecodingResult?.language
+                }
 
                 decodingResult = try await textDecoder.decodeText(
                     from: encoderOutput,
                     using: decoderInputs,
                     sampler: tokenSampler,
-                    options: options,
+                    options: currentDecodingOptions,
```
Comment (ZachNagengast) [inline Sources/WhisperKit/Core/WhisperKit.swift:667]: Is this renaming unnecessary now?

**Review by ZachNagengast — CHANGES_REQUESTED** [review]
Great progress, just need those last tests and cleanup and its good to go!

File: Sources/WhisperKit/Core/LogitsFilter.swift (hunk 2 above)
Comment (Abhinay1997) [reply Sources/WhisperKit/Core/LogitsFilter.swift]: This is done

File: Tests/WhisperKitTests/UnitTests.swift (hunk 4 above)
Comment (Abhinay1997) [reply Tests/WhisperKitTests/UnitTests.swift:577]: Yes Zach, this is only for the logits filter test. The language detection test will be added near the tests for the other languages

File: Sources/WhisperKit/Core/TextDecoder.swift (hunk 3 above)
Comment (Abhinay1997) [reply Sources/WhisperKit/Core/TextDecoder.swift:411]: Let me do it in a seperate PR ?

File: Sources/WhisperKit/Core/WhisperKit.swift (hunk 5 above)
Comment (Abhinay1997) [reply Sources/WhisperKit/Core/WhisperKit.swift:667]: It would be needed as I have to pass the detected language to the decoder loop so it can skip the forward pass on language token.

File: Sources/WhisperKit/Core/TextDecoder.swift (hunk 3 above)
Comment (ZachNagengast) [reply Sources/WhisperKit/Core/TextDecoder.swift:411]: Sure 👍
